### ISSUE TITLE
Error 139 when trying to exec supplied cwebp

### ISSUE THREAD

**Issue by mrbig00** [rosell-dk/webp-convert#92]
I know it's not really related to the project itself, but did you guys ever encountered this error? Any ideas how to solve it?

**Comment by rosell-dk** [comment]
I found this answer [on the web](https://groups.google.com/forum/#!topic/omnetpp/_zonGWenUqg): "Unix systems return errono 128+signal when a signal received. 128 + 11 = 139 . SIgnal 11 is SIGSEV (i.e. segmentation violation). = There is a memory access bug in your C++ code."

My guess is that the prebuild binary for your OS doesn't work on your system. It might even be corrupt

**Comment by rosell-dk** [comment]
You can try to replace the binary in the plugin with another binary.

Prebuild binaries can be downloaded here: https://storage.googleapis.com/downloads.webmproject.org/releases/webp/index.html

Download the zip, and copy the *bin/cwebp* file to the plugin (*wp-content/plugins/webp-express/vendor/rosell-dk/webp-convert/src/Converters/Binaries/*)

There are instructions for compiling from source here:
https://developers.google.com/speed/webp/docs/using

**Comment by rosell-dk** [comment]
In the cwebp options on WebP Express options page, you can configure where to look for cwebp (Try to execute cweb binary at common locations  / Try precompiled cwebp). Just to be sure that the problem is with a precompiled cwebp, can you try disabling "Try precompiled cwebp", and tell me if the 139 error persists?

**Comment by rosell-dk** [comment]
The binaries is btw copied from ewww plugin a while back. I guess they should be updated...

**Comment by rosell-dk** [comment]
I should also mention that I am considering an option to simply run 'cwebp' on the system, without pointing to a specific binary. cwebp might be defined in the path.

**Comment by rosell-dk** [comment]
Oh, one more thing.
The plugin verifies if the executable has a known hash, before executing

These are defined in vendor/src/Converters/Cwebp.php

```
    // OS-specific binaries included in this library, along with hashes
    private static $suppliedBinariesInfo = [
        'WinNT' => [ 'cwebp.exe', '49e9cb98db30bfa27936933e6fd94d407e0386802cb192800d9fd824f6476873'],
        'Darwin' => [ 'cwebp-mac12', 'a06a3ee436e375c89dbc1b0b2e8bd7729a55139ae072ed3f7bd2e07de0ebb379'],
        'SunOS' => [ 'cwebp-sol', '1febaffbb18e52dc2c524cda9eefd00c6db95bc388732868999c0f48deb73b4f'],
        'FreeBSD' => [ 'cwebp-fbsd', 'e5cbea11c97fadffe221fdf57c093c19af2737e4bbd2cb3cd5e908de64286573'],
        'Linux' => [ 'cwebp-linux', '916623e5e9183237c851374d969aebdb96e0edc0692ab7937b95ea67dc3b2568']
    ];
```
You will have to update the hash string.
Or, if you are lazy, simply change line 314 in Cwebp.php

from: `if (function_exists('hash_file')) {`
to: `if (false)`

You actually have to find the same line in */vendor/build/webp-on-demand-2.inc*.
The src file is used when clicking "test"
The build file is used when images are being redirected to the PHP

**Comment by mrbig00** [comment]
made my day

> Or, if you are lazy, simply change line 314 in Cwebp.php

I actually already replaced the binary (and the hash) with the most recent one from the source you mentioned, but it's the same error. I think that the problem could be some incompatible stuff between the compiled cwebp and the system. 

Maybe we could implement some meaningful Exception based on the return codes, what do you think?

**Comment by rosell-dk** [comment]
Sure - the more information, the better. Just haven't found the big cheat-sheet of return codes yet :)
But 139 can now get a meaningfull message.

**Comment by rosell-dk** [comment]
I guess all of them in layman terms means "The binary did not work". So I will show that message for errorCode > 128 and point to this issue

**Comment by mrbig00** [comment]
Regarding this issue, I was thinking that a new option for the \WebPConvert\Converters\Cwebp converter where you can define your own binary path in cases like this, so it's easier to debug. What do you think? 

I'm happy to implement it around next week.

**Comment by rosell-dk** [comment]
I have myself pondered doing this, but I am worried about the security implications of doing it

**Comment by rosell-dk** [comment]
In the upcoming 2.0 (any day now), I allow sensitive options to be set with environment variables (ie the api key for ewww). I could allow the binary path to be set like that.

**Comment by rosell-dk** [comment]
The path can be set with the environment variable CWEBP_PATH

To set environment variables in Apache, you can add a line like this in your `.htaccess` or vhost configuration:
```
SetEnv CWEBP_PATH /usr/bin/cwebp
```

To find out where the cwebp binary is, open a console and type:
```
whereis cwebp
```

**Comment by rosell-dk** [comment]
NOTE NOTE NOTE:
I have changed the name of the environment variable to "WEBPCONVERT_CWEBP_PATH" in the last commit (will be released as 2.0.3). I'm aware that this is an API change and it should trigger a major version shift, - but I am not doing that, hoping that only few has yet migrated to 2.0 and only few are using this feauture

Btw, you can now also set it with a contant:
```php
define('WEBPCONVERT_CWEBP_PATH', '/usr/bin/cwebp');
```
